### PR TITLE
Add Gemma-4 multimodal image input for vLLM

### PR DESCRIPTION
**Pull request by sshonTT**
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-xla/issues/4702)

### Problem description
There's not yet support for multimodal model in vLLM

### What's changed
Enable the multimodal encoder path in the TT model runner: activate supports_mm_inputs, execute/gather the mm encoder (handle the is_embed mask for models that interleave marker tokens), precompile embed_input_ids, and build the dummy mm batch via the current vLLM API.

overrides: route HF vision/audio tower attention through the fused TT SDPA op (eager attention can't lower over the device mesh); replace vLLM's masked_scatter_-based _merge_multimodal_embeddings with a static-shape cumsum+gather+where impl (the dynamic mask scatter is rejected by tt-mlir's Shardy pass); promote pre-allocated plain-tensor attrs to non-persistent  buffers so model.to(device) relocates them.

ascend_scheduler: schedule multimodal encoder inputs (the base scheduler does this inline with KV allocation; do it post-hoc to stay close to AscendScheduler.schedule()).

Add a BHQB 31B image-input test.

### Checklist
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: integrations/vllm_plugin/vllm_tt/overrides.py [hunk 1]
```diff
@@ -15,6 +16,95 @@
...
+
+    if pad_q > 0:
+        attn_output = attn_output[:, :, :seq_q, :]
+
+    attn_output = attn_output.transpose(1, 2).contiguous()
+    return attn_output, None
+
+
+# Idempotently register so reloads (e.g. pytest reruns) don't raise.
+if "tt" not in ALL_ATTENTION_FUNCTIONS:
```
Comment (AleksKnezevic) [inline integrations/vllm_plugin/vllm_tt/overrides.py]: what's this?

File: integrations/vllm_plugin/vllm_tt/overrides.py [hunk 2]
```diff
@@ -168,6 +258,90 @@ def tt_rotary_embedding_module(layer: torch.nn.Module) -> torch.nn.Module:
...
+    logger.info(
+        "Installed static-shape _merge_multimodal_embeddings (replaces "
+        "masked_scatter_-based default that emits dynamic shapes)."
+    )
+
+
+_install_static_shape_merge_multimodal_embeddings()
+
+
+def _promote_pre_allocated_attrs_to_buffers(model: torch.nn.Module) -> None:
```
Comment (AleksKnezevic) [inline integrations/vllm_plugin/vllm_tt/overrides.py:41]: why is this needed?

File: integrations/vllm_plugin/vllm_tt/overrides.py [hunk 3]
```diff
@@ -168,6 +258,90 @@ def tt_rotary_embedding_module(layer: torch.nn.Module) -> torch.nn.Module:
 ]
 
 
+def _route_vision_attention_to_tt(model: torch.nn.Module) -> None:
```
Comment (AleksKnezevic) [inline integrations/vllm_plugin/vllm_tt/overrides.py]: why is this needed?

File: integrations/vllm_plugin/vllm_tt/overrides.py (hunk 1 above)
Comment (sshonTT) [reply integrations/vllm_plugin/vllm_tt/overrides.py]: The Gemma4 vision tower is instantiated as a plain HF module (AutoModel.from_config) and run eager attention. vLLM has no attention path for it (unlike the language model, which goes through vLLM's attention backend). 
So to make its attention use our SDPA, we register the op into HF's ALL_ATTENTION_FUNCTIONS registry under "tt", and _route_vision_attention_to_tt points the tower's _attn_implementation at it.

File: integrations/vllm_plugin/vllm_tt/overrides.py (hunk 3 above)
Comment (sshonTT) [reply integrations/vllm_plugin/vllm_tt/overrides.py]: Please see this [one](https://github.com/tenstorrent/tt-xla/pull/5029#discussion_r3335521893)

File: integrations/vllm_plugin/vllm_tt/overrides.py (hunk 2 above)
Comment (sshonTT) [reply integrations/vllm_plugin/vllm_tt/overrides.py:41]: We don't need this right now. Will remove it. 
For quick explanation, it's for e4b model which has per layer embeddings which gemma4 model pre-allocates as a plain tensor attribute, not a registered buffer. 
model.to(device) only moves parameters and registered buffers, so it's stranded on CPU and a later op mixes it with on-device tensors, so dynamo fails on mixed-device fake tensors. Re-registering it as a non-persistent buffer makes .to(device) follow it.

Comment (sshonTT) [reply integrations/vllm_plugin/vllm_tt/overrides.py:41]: Actually, we need it rn as I've added e4b test to PR. I had to debug kv sharing feature in e4b and now I found a fix.

**Comment by sshonTT** [comment]
nightly passes : https://github.com/tenstorrent/tt-xla/actions/runs/26772188227

**Comment by sshonTT** [comment]
@kmabeeTT, @mmanzoorTT could you ptal at this PR? Thanks!

File: integrations/vllm_plugin/vllm_tt/overrides.py [hunk 4]
```diff
@@ -17,6 +18,94 @@
 logger = tt_init_logger(__name__)
 
 
+def tt_sdpa_attention_forward(
```
Comment (mmanzoorTT) [inline integrations/vllm_plugin/vllm_tt/overrides.py]: Suggestion: Move this definition to separate file and use `overrides.py` to just calling override function like we are doing for other layers.

Comment (sshonTT) [reply integrations/vllm_plugin/vllm_tt/overrides.py]: That's a fair point. I've moved the definition. Thanks!

File: tests/integrations/vllm_plugin/generative/test_gemma4_generation.py [hunk 5]
```diff
@@ -70,3 +74,104 @@ def test_generation_bhqb_gemma4_26b_a4b():
...
+        "model": model_name,
+        "limit_mm_per_prompt": {"image": 1, "video": 0, "audio": 0},
+        "max_num_batched_tokens": 2560,
+        "max_num_seqs": 1,
+        "max_model_len": 1024,
+        "gpu_memory_utilization": 0.1,
+        "additional_config": {
+            "enable_const_eval": True,
+            "min_context_len": 32,
+            "enable_tensor_parallel": True,
```
Comment (mmanzoorTT) [inline tests/integrations/vllm_plugin/generative/test_gemma4_generation.py:171]: It is using `enable_tensor_parallel=True` but not marked with `@pytest.mark.tensor_parallel`; is this intentional?

File: tests/integrations/vllm_plugin/generative/test_gemma4_generation.py [hunk 6]
```diff
@@ -70,3 +74,104 @@ def test_generation_bhqb_gemma4_26b_a4b():
...
+    output_text = llm.chat(messages, sampling_params)[0].outputs[0].text
+    print(f"output: {output_text}")
+    assert_output_coherent(output_text)
+
+    check_host_memory(model_name)
+
+
+@pytest.mark.nightly
+@pytest.mark.bhqb
+def test_generation_bhqb_gemma4_26b_a4b_image():
```
Comment (mmanzoorTT) [inline tests/integrations/vllm_plugin/generative/test_gemma4_generation.py]: I suggest to move `test_tensor_parallel_generation_bhqb_gemma4_31b_image` here and parametrize it with `test_generation_bhqb_gemma4_26b_a4b_image`. The only difference is the model being used and different mesh shape (1D vs 2D mesh).

**Review by mmanzoorTT — APPROVED** [review]

File: tests/integrations/vllm_plugin/generative/test_gemma4_generation.py (hunk 5 above)
Comment (sshonTT) [reply tests/integrations/vllm_plugin/generative/test_gemma4_generation.py:171]: Ah missed this. Thanks!

File: tests/integrations/vllm_plugin/generative/test_gemma4_generation.py (hunk 6 above)
Comment (sshonTT) [reply tests/integrations/vllm_plugin/generative/test_gemma4_generation.py]: Yeah that makes sense. Moved the test

**Review by mmanzoorTT — APPROVED** [review]
